### PR TITLE
background - introduce observable store for configManager state

### PR DESCRIPTION
**Pull request by kumavis**
Updates #979
Updates #847
Updates #850

### CODE REVIEW

**Review by danfinlay — CHANGES_REQUESTED** [review]
Making a few changes, moving observable store into metamask controller.

**Review by danfinlay — APPROVED** [review]
I'm happy with this if the tests pass now. It will make metamask slightly more performant, and make it much easier to convert to async storage later on.

**Review by danfinlay — CHANGES_REQUESTED** [review]
Oh, thought of another way to change it:  The `loadData` method passed to `configManager` needs to pull from the observable store, not the options method, which can probably be removed.

I'll do it after lunch right now...

**Comment by kumavis** [comment]
maybe better to set the `initState` from the outside after instantiating the `metamaskController` to make sure that works and is not awkward, as we will need to be able to do that from `background.js` when we move to async persistence

File: app/scripts/metamask-controller.js [hunk 1]
```diff
@@ -14,14 +14,24 @@ const autoFaucet = require('./lib/auto-faucet')
...
-    this.state = { network: 'loading' }
     this.opts = opts
+    this.state = { network: 'loading' }
+
+    // Observable wrapper around persistence strategy:
+    this.configStore = new ObservableStore(opts.initState)
+    this.configStore.subscribe(opts.setData)
+
     this.configManager = new ConfigManager(opts)
+    this.configManager.subscribe((state) => {
```
Comment (kumavis) [inline app/scripts/metamask-controller.js]: we will likely be replacing  `this.configManager.subscribe` with ` this.configManager.store.subscribe`, so I think doing it this way further complicates things

**Comment by flyswatter** [comment]
Why wouldn't we just say "load initial state before starting MetaMask" is a requirement? That's the current situation, and it could also work with async loading.

**Comment by kumavis** [comment]
created a new PR where I am experimenting with the pattern
https://github.com/MetaMask/metamask-plugin/pull/999

**Comment by kumavis** [comment]
This one's going back to the drawing board a bit

**Pull request closed without merging**